### PR TITLE
board: limifrog-v1 initial commit

### PR DESCRIPTION
**Pull request by katezilla**

### CODE REVIEW

**Comment by katezilla** [comment]
board: Initial Commit for LimiFrog Version 1
Original Issue: https://github.com/RIOT-OS/RIOT/issues/3020#issuecomment-112323596

File: boards/limifrog-v1/Makefile.features [hunk 1]
```diff
@@ -0,0 +1,7 @@
+EATURES_PROVIDED += periph_gpio
```
Comment (PeterKietzmann) [inline boards/limifrog-v1/Makefile.features]: **F**EATURES

File: boards/limifrog-v1/board.c [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ *
+ * The LED initialization is hard-coded in this function.
+ * As the LED is soldered onto the board it is fixed to
+ * its CPU pins.
+ *
+ * The red LED is connected to pin PC3
+ */
+static void leds_init(void)
+{
+    /* enable clock for port GPIOA */
```
Comment (PeterKietzmann) [inline boards/limifrog-v1/board.c]: I think it must be GPIO**C** if the LED is connected to PC3 (like stated in line 41)

File: boards/limifrog-v1/board.c [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    LED_RED_PORT->OSPEEDR |= GPIO_OSPEEDER_OSPEEDR3;
+    /* set output type to push-pull */
+    LED_RED_PORT->OTYPER &= ~(GPIO_OTYPER_OT_3);
+    /* configure pin as general output */
+    LED_RED_PORT->MODER &= ~(GPIO_MODER_MODER3);
+    LED_RED_PORT->MODER |= GPIO_MODER_MODER3_0;
+    /* disable pull resistors */
+    LED_RED_PORT->PUPDR &= ~(GPIO_PUPDR_PUPDR3);
+    /* turn all LEDs off */
+    LED_RED_PORT->BRR = 1 << 3;
```
Comment (PeterKietzmann) [inline boards/limifrog-v1/board.c]: could you put parenthesis around `1 << 3` please

File: boards/limifrog-v1/dist/openocd.cfg [hunk 4]
```diff
@@ -0,0 +1,5 @@
+source [find interface/stlink-v2.cfg]
+transport select hla_swd
+
+set WORKAREASIZE 0x4000
```
Comment (PeterKietzmann) [inline boards/limifrog-v1/dist/openocd.cfg]: Where does this value come from? I thought `0x4000 = 0d16384` which stands for 16 kB RAM (thats just what I thought), but the `stm32l1` hast 80kB of RAM.

Comment (PeterKietzmann) [inline boards/limifrog-v1/dist/openocd.cfg]: In the regarding file I found:
```
# Work-area is a space in RAM used for flash programming
# By default use 10kB
if { [info exists WORKAREASIZE] } {
   set _WORKAREASIZE $WORKAREASIZE
} else {
   set _WORKAREASIZE 0x2800
}
```
So forget about the 80k above and let's initially go with `0x2800` :-)

File: boards/limifrog-v1/include/board.h [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+/**
+ * @name Macros for controlling the on-board LEDs.
+ * @{
+ */
+#define LED_GREEN_ON
+#define LED_GREEN_OFF
+#define LED_GREEN_TOGGLE
+#define LED_RED_ON          (LED_RED_PORT->BSSRL = (1 << LED_RED_PIN))
+#define LED_RED_OFF         (LED_RED_PORT->BSSRH = (1 << LED_RED_PIN))
+#define LED_RED_TOGGLE      (LED_RED_PORT->ODR ^= (1 << LED_RED_PIN))
```
Comment (PeterKietzmann) [inline boards/limifrog-v1/include/board.h]: Please add 
```
#define LED_ORANGE_ON
#define LED_ORANGE_OFF
#define LED_ORANGE_TOGGLE
```
even if it's not available

File: boards/limifrog-v1/include/periph_conf.h [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @name Clock system configuration
+ * @{
+ **/
+#define CLOCK_HSE           (37000U)            /* external oscillator */
```
Comment (PeterKietzmann) [inline boards/limifrog-v1/include/periph_conf.h]: Did you find this information in "frog-sheets"?

File: boards/limifrog-v1/include/periph_conf.h [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+ */
+#define TIMER_NUMOF         (1U)
+#define TIMER_0_EN          1
+#define TIMER_IRQ_PRIO      1
+
+/* Timer 0 configuration */
+#define TIMER_0_DEV_0       TIM2
+#define TIMER_0_DEV_1       TIM3
+#define TIMER_0_CHANNELS    4
+#define TIMER_0_PRESCALER   (16U)
```
Comment (PeterKietzmann) [inline boards/limifrog-v1/include/periph_conf.h]: I didn't look this up. Just wondering why it is `32U` on the nucleo-l1 board

File: boards/limifrog-v1/include/periph_conf.h [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+#define TIMER_0_IRQ_CHAN_1  TIM3_IRQn
+#define TIMER_0_IRQ_PRIO    1
+#define TIMER_0_TRIG_SEL    TIM_SMCR_TS_0
+/** @} */
+
+/**
+ * @brief UART configuration
+ * @{
+ */
+#define UART_NUMOF          (1U)
```
Comment (PeterKietzmann) [inline boards/limifrog-v1/include/periph_conf.h]: IIRC there is one sensor (or was it the BLE device) which is connected to on of the three available `UART`s of the MCU. Could you add that one and document it? (Just if I'm not wrong in my assumption)

File: boards/limifrog-v1/include/periph_conf.h [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+#define UART_0_RX_PIN       11
+#define UART_0_TX_PIN       10
+#define UART_0_AF           7
+/** @} */
+
+/**
+ * @brief SPI configuration
+ * @{
+ */
+#define SPI_NUMOF           (1U)
```
Comment (PeterKietzmann) [inline boards/limifrog-v1/include/periph_conf.h]: I don't know out of my head but **if** the display is internally connected to one SPI bus, please add the regarding bus here and document it

File: boards/limifrog-v1/include/periph_conf.h [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+#define SPI_0_PIN_MOSI      7
+#define SPI_0_PIN_MISO      6
+#define SPI_0_PIN_AF        5
+/** @} */
+
+/**
+ * @name I2C configuration
+  * @{
+ */
+#define I2C_NUMOF           (1U)
```
Comment (PeterKietzmann) [inline boards/limifrog-v1/include/periph_conf.h]: Same as for the `UART` and `I2C` ports. There are some sensors that are connected to one of the available `I2C` buses. Please add the right one here

File: boards/limifrog-v1/include/periph_conf.h [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+#define GPIO_2_CLKEN()      (RCC->AHBENR |= RCC_AHBENR_GPIOCEN)
+#define GPIO_2_EXTI_CFG()   (SYSCFG->EXTICR[3] |= SYSCFG_EXTICR4_EXTI12_PC)
+#define GPIO_2_IRQ          EXTI15_10_IRQn
+/* GPIO channel 3 config */
+#define GPIO_3_PORT         GPIOB
+#define GPIO_3_PIN          0
+#define GPIO_3_CLKEN()      (RCC->AHBENR |= RCC_AHBENR_GPIOAEN)
+#define GPIO_3_EXTI_CFG()   (SYSCFG->EXTICR[0] |= SYSCFG_EXTICR1_EXTI0_PB)
+#define GPIO_3_IRQ          EXTI0_IRQn
+
```
Comment (PeterKietzmann) [inline boards/limifrog-v1/include/periph_conf.h]: In  #3095 we slightly changed the gpio interface and implementation "style". As the stml1 and the nucleo-l1 board are still in the old fashion, I think it's ok to go this way. In a second step (PR) I'd recommend to adapt both ;-)

Comment (kaspar030) [inline boards/limifrog-v1/include/periph_conf.h]: Actually I would would vote against merging with use of that legacy device.
Please adapt, it's not that hard.

File: cpu/stm32l1/ldscripts/stm32l151rc.ld [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+ * @author          Katja Kirstein <katja.kirstein@haw-hamburg.de>
+ *
+ * @}
+ */
+
+MEMORY
+{
+    rom (rx)        : ORIGIN = 0x08000000, LENGTH = 256K
+    ram (xrw)       : ORIGIN = 0x20000000, LENGTH = 32K
+}
```
Comment (PeterKietzmann) [inline cpu/stm32l1/ldscripts/stm32l151rc.ld]: Did you check these values? I thought the memory is bigger...

File: cpu/stm32l1/ldscripts/stm32l151rc.ld [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * @addtogroup      cpu_stm32l1
+ * @{
+ *
+ * @file
+ * @brief           Memory definitions for the STM32L151RC
+ *
+ * @author          Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author          Katja Kirstein <katja.kirstein@haw-hamburg.de>
```
Comment (PeterKietzmann) [inline cpu/stm32l1/ldscripts/stm32l151rc.ld]: Some days ago I discussed with Hauke, that it's fine to replace the author if a file is just used as a "skeleton" or "template". Here I think you can apply these outcomes :-) . (Especially as the code is just 2 lines long)

**Comment by PeterKietzmann** [comment]
@katezilla thanks for this PR! There are some open questions which must not be faults, but just my unknowing of details. But in general the PR is in a good shape :-)

File: cpu/stm32l1/ldscripts/stm32l151rc.ld (hunk 12 above)
Comment (katezilla) [inline cpu/stm32l1/ldscripts/stm32l151rc.ld]: yes i did. Datasheet says: 256KB Flash, 32KB SRAM,

File: boards/limifrog-v1/include/periph_conf.h (hunk 11 above)
Comment (haukepetersen) [inline boards/limifrog-v1/include/periph_conf.h]: +1 here. Adapting the gpio code for the STM32L1 is pretty much straight forward, just have a look at the `stm32f4` gpio code... This way we would again have one more CPU moved to a cleaner and nicer GPIO implementation...

File: boards/limifrog-v1/include/periph_conf.h [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+/**
+ * @brief SPI configuration
+ * @{
+ */
+#define SPI_NUMOF           (2U)
+#define SPI_0_EN            1
+#define SPI_1_EN            1
+
+/* SPI 0 device configuration */
+#define SPI_0_DEV           SPI1  /* Used for OLED Display and external pins */
```
Comment (PeterKietzmann) [inline boards/limifrog-v1/include/periph_conf.h]: Sorry for being so picky. Could you include the model name here? E.G. " *Densitron DD-160128FC-1a OLED display; external pins* "

File: boards/limifrog-v1/include/periph_conf.h [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+/* SPI 0 pin configuration */
+#define SPI_0_PORT_CLKEN()  (RCC->AHBENR |= RCC_AHBENR_GPIOAEN)
+#define SPI_0_PORT          GPIOA
+#define SPI_0_PIN_SCK       5
+#define SPI_0_PIN_MOSI      7
+#define SPI_0_PIN_MISO      6
+#define SPI_0_PIN_AF        5
+
+/* SPI 1 device configuration */
+#define SPI_1_DEV           SPI3          /* Adesto AT45DB641E */
```
Comment (PeterKietzmann) [inline boards/limifrog-v1/include/periph_conf.h]: Maybe " *Adesto AT45DB641E data flash* "

File: boards/limifrog-v1/include/periph_conf.h [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+/* I2C 0 pin configuration */
+#define I2C_0_PORT_CLKEN()  (RCC->AHBENR |= RCC_AHBENR_GPIOBEN)
+#define I2C_0_PORT          GPIOB
+#define I2C_0_SCL_PIN       8
+#define I2C_0_SCL_AF        4
+#define I2C_0_SDA_PIN       9
+#define I2C_0_SDA_AF        4
+
+/* I2C 1 device configuration */
+#define I2C_1_DEV           I2C2         /* Used for Sensors */
```
Comment (PeterKietzmann) [inline boards/limifrog-v1/include/periph_conf.h]: Could you list the sensors? For me it's fine to do it without a description here. So: "*ST VL6180X, ST LSM6DS3, ST LIS3MDL, ST SLPS25H*"

File: boards/limifrog-v1/include/periph_conf.h [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+#define UART_0_ISR          isr_usart3
+/* UART 0 pin configuration */
+#define UART_0_PORT         GPIOC
+#define UART_0_PORT_CLKEN() (RCC->AHBENR |= RCC_AHBENR_GPIOCEN)
+#define UART_0_RX_PIN       11
+#define UART_0_TX_PIN       10
+#define UART_0_AF           7
+
+/* UART 1 device configuration */
+#define UART_1_DEV          USART3        /* BTLE */
```
Comment (PeterKietzmann) [inline boards/limifrog-v1/include/periph_conf.h]: I prefer the previous description: *"Panasonic PAN1740 BLE module"*

File: boards/limifrog-v1/include/periph_conf.h (hunk 11 above)
Comment (PeterKietzmann) [inline boards/limifrog-v1/include/periph_conf.h]: @katezilla what do you think, can you do it :-) ?

**Comment by PeterKietzmann** [comment]
@katezilla thanks for addressing some of my comments. Your PR looks great now! There is still the thing about the gpio mapping. Could you please adapt that? 
Did you already test some peripherals? In that case, I'd ask for lending me the hardware.

File: boards/limifrog-v1/include/periph_conf.h (hunk 11 above)
Comment (katezilla) [inline boards/limifrog-v1/include/periph_conf.h]: sure, i can do that

File: boards/limifrog-v1/include/board.h [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+ */
+#define HW_TIMER            TIMER_0
+
+/**
+ * @name LED pin definitions
+ * @{
+ */
+#define LED_RED_PORT      (GPIOC)
+#define LED_RED_PIN       (3)
+/** @} */
```
Comment (PeterKietzmann) [inline boards/limifrog-v1/include/board.h]: Could you add `#define LED_RED_GPIO        GPIO(PORT_C,3)` here please?

File: boards/limifrog-v1/include/board.h [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+ */
+void board_init(void);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* BOARD_H_ */
+/** @} */
+/** @} */
```
Comment (PeterKietzmann) [inline boards/limifrog-v1/include/board.h]: I think you can remove one of these lines, l right?

File: cpu/stm32l1/periph/gpio.c [hunk 20]
```diff
@@ -24,545 +26,196 @@
...
-
-    pin = gpio_pin_map[dev];
-    EXTI->IMR |= (1 << pin);
+    /* set pin to AF mode */
+    port->MODER &= ~(3 << (2 * pin_num));
+    port->MODER |= (2 << (2 * pin_num));
+    /* set selected function */
+    port->AFR[(pin_num > 7) ? 1 : 0] &= ~(0xf << ((pin_num & 0x07) * 4));
+    port->AFR[(pin_num > 7) ? 1 : 0] |= (af << ((pin_num & 0x07) * 4));
 }
```
Comment (PeterKietzmann) [inline cpu/stm32l1/periph/gpio.c]: I forgot the exact reason but IIRC @haukepetersen decided to enable the clock for the selected port here as well.

File: cpu/stm32l1/periph/gpio.c [hunk 21]
```diff
@@ -24,545 +26,196 @@
...
+int gpio_read(gpio_t pin)
+{
+    GPIO_TypeDef *port = _port(pin);
+    uint32_t pin_num = _pin_num(pin);
 
-    if (port->MODER & (1 << (pin * 2))) {       /* if configured as output */
-        return port->ODR & (1 << pin);          /* read output data register */
+    if (port->MODER & (3 << (pin_num * 2))) {   /* if configured as output */
+        return port->ODR & (1 << pin_num);      /* read output data reg */
     } else {
```
Comment (PeterKietzmann) [inline cpu/stm32l1/periph/gpio.c]: According to our coding conventions, could you put the else branch in a new line please.

File: cpu/stm32l1/periph/i2c.c [hunk 22]
```diff
@@ -486,6 +496,13 @@ void i2c_poweroff(i2c_t dev)
             I2C_0_CLKDIS();
             break;
 #endif
+#if I2C_1_EN
+        case I2C_1:
+            while (I2C_0_DEV->SR2 & I2C_SR2_BUSY);
+
```
Comment (PeterKietzmann) [inline cpu/stm32l1/periph/i2c.c]: I'm wondering about these newlines (here and in line 495). I know, it's not your fault ;-)

**Comment by PeterKietzmann** [comment]
@katezilla great work! I think we are close before merging when the minor comments from today are addressed. 
For the protocol: I successfully tested SPI and I2C yesterday. Will check GPIO now

**Comment by PeterKietzmann** [comment]
Using GPIOs as input and output works. The interrupt handler seems to cause a hard fault. Before spending too much time in searching the error, could someone with a nucleo-l1 board please check if this appears on that board with and without this PR (that also changes cpu/stml1/periph/gpio.c)? Thinking about @haukepetersen ;-) , or who else owns this board?

**Comment by OlegHahm** [comment]
> @katezilla great work! I think we are close before merging when the minor comments from today are addressed.
> For the protocol: I successfully tested SPI and I2C yesterday. Will check GPIO now

Very nice indeed! Looking forward to see the Limifrog with RIOT on it in action.

**Comment by OlegHahm** [comment]
P.S. I own an Nucleo L1, too.

**Comment by PeterKietzmann** [comment]
Oh, good to know :-) ! If you find a minute you could try with the periph_gpio example, if the interrupt handler causes a hard fault without and with this PR.

**Comment by OlegHahm** [comment]
I don't have it with me right now. Can you ping me on Monday?

**Comment by PeterKietzmann** [comment]
If Hauke haven't checked it until then, I'll do so.

**Comment by PeterKietzmann** [comment]
@OlegHahm, @haukepetersen this is a reminder to test above described behaviour.

**Comment by OlegHahm** [comment]
I'm on it.

**Comment by haukepetersen** [comment]
Hej, just looked at the GPIO driver together with @OlegHahm. With our board here, even in- and output does not seem to work with the gpio test app anymore (but this might be caused by the legacy implementation not wanting to work with the test case...). I think debugging the GPIO driver in it's current state does not make too much sense, since it is old legacy code. The nicer option would be to re-implement the GPIO driver in a clean way (as done for the f1/f3/f4).

**Comment by PeterKietzmann** [comment]
Hm, looking just roughly over the driver, it does not look too bad or different from the other ones that you mentioned (for me). Anyway, would you agree that a new driver implementation is out of scope of this PR? What do you think about using the compat-header-solution in this PR and working on a follow-up PR regarding the driver implementation? If it is @katezilla who will do this, it would be nice if you could lend us your nucleo-l1.

**Comment by PeterKietzmann** [comment]
You need to adapt the vector table in cpu/stm32f1/vectors.c . Please compare how it was done in #3535.

**Comment by PeterKietzmann** [comment]
Works :-) . Please squash your commits into two. One should hold changes about the stm32l1 changes (gpio driver and interrupt vector). The other one should contain the porting of the limifrog board. Note that the linkerscript is located in the CPU folder but belongs to the porting of the board. It does not matter in which order the commits apper

**Comment by katezilla** [comment]
squashed

**Comment by PeterKietzmann** [comment]
Dammit! Travis complains about some applications. You should add `#include "cpu.h" ` in [periph.cpu:23](https://github.com/RIOT-OS/RIOT/pull/3509/files#diff-413da689e19af3a4acca7a23bb8dffa0R23)

**Comment by PeterKietzmann** [comment]
Thanks! Next time please write a short line that comments addressed. Otherwise I'll just see this randomly.

**Comment by katezilla** [comment]
added the include :)

**Comment by PeterKietzmann** [comment]
Happy times, and go

**Pull request merged**